### PR TITLE
Allow different energy binning for true and reconstructed energy in Maps

### PR DESCRIPTION
**Pull request by AtreyeeS**
This PR lets maps have different geometries for the true and reco energy.
The exposure and PSF map is in true energies, the counts and background map in reco energies.
Changes are made in map creation (MapMaker) and model evaluation and fitting.

### CODE REVIEW

File: gammapy/cube/fit.py [hunk 1]
```diff
@@ -208,22 +209,37 @@ def apply_psf(self, npred):
         """Convolve npred cube with PSF"""
         return npred.convolve(self.psf)
 
-    def apply_edisp(self, data):
-        """Convolve map data with energy dispersion."""
-        data = np.rollaxis(data, 0, 3)
+    def apply_edisp(self, npred):
+        """Convolve map data with energy dispersion.
+
+        npred: npred map in e_true
```
Comment (cdeil) [inline gammapy/cube/fit.py]: Fix docstring formatting

File: gammapy/cube/fit.py [hunk 2]
```diff
@@ -208,22 +209,37 @@ def apply_psf(self, npred):
...
+        Returns
+        -------
+        npred_reco: npred map in e_reco
+        """
+
+        loc = npred.geom.get_axes_dimension_by_name("energy")
+        data = np.moveaxis(npred.data, loc, -1)
```
Comment (cdeil) [inline gammapy/cube/fit.py]: Use numpy rollaxis, we still support old Numpy versions.

File: gammapy/cube/fit.py [hunk 3]
```diff
@@ -208,22 +209,37 @@ def apply_psf(self, npred):
...
         data = np.dot(data, self.edisp.pdf_matrix)
-        return np.rollaxis(data, 2, 0)
+        data1 = np.moveaxis(data, -1, loc) #now dim is in ereco
```
Comment (cdeil) [inline gammapy/cube/fit.py]: You are modifying `data` and still keep calling the variable `data` on the lines above.
Suggest to continue here and not start calling it `data1`.

File: gammapy/cube/fit.py [hunk 4]
```diff
@@ -208,22 +209,37 @@ def apply_psf(self, npred):
...
+            array of the predicted counts in each bin (in reco energy)
         """
         flux = self.compute_flux()
         npred = self.apply_exposure(flux)
         if self.psf is not None:
             npred = self.apply_psf(npred)
-        # TODO: discuss and decide whether we need to make map objects in `apply_aeff` and `apply_psf`.
         if self.edisp is not None:
-            npred.data = self.apply_edisp(npred.data)
+            npred = self.apply_edisp(npred)
```
Comment (cdeil) [inline gammapy/cube/fit.py]: I think that's a good change, to always return Map objects now.
If this is a significant inefficiency we can change again later, but probably it's nicer / cleaner like this.

File: gammapy/cube/fit.py [hunk 5]
```diff
@@ -208,22 +209,37 @@ def apply_psf(self, npred):
...
         if self.background:
             npred.data += self.background.data
         return npred.data
```
[the remark this answers is not in the document]
Comment (AtreyeeS) [reply gammapy/cube/fit.py]: This is breaking the simulation notebooks. The MapFit only needs a `~numpy.ndarray` to compute the statistics, so not changing it now.

File: gammapy/maps/geom.py [hunk 6]
```diff
@@ -1395,6 +1395,23 @@ def get_axis_by_name(self, name):
...
+           Name of the requested axis
+
+        Returns
+        -------
+        i : integer
+            The number of the dimension of the specified axis in the map geom
+
+        """
+        axes = {axis.name.upper(): idx for idx, axis in enumerate(self.axes)}
+        return axes[name.upper()]
```
Comment (registerrier) [inline gammapy/maps/geom.py]: you should return an index not the axis here, no?

```
names = [axis.name for axis in self.axes]
return names.index(name)
```

File: gammapy/maps/geom.py [hunk 7]
```diff
@@ -1421,3 +1421,26 @@ def copy(self, **kwargs):
             Copied map geometry.
         """
         return self._init_copy(**kwargs)
+
+    def get_axes_dimension_by_name(self, name):
```
[the remark this answers is not in the document]
Comment (AtreyeeS) [reply gammapy/maps/geom.py]: I have put the test in test_wcs, with a `raises(ValueError)`

**Pull request merged**